### ISSUE TITLE
Hermes plugin

### ISSUE THREAD

**Issue by daniele-athome** [AppDaemon/appdaemon#989]
[Over at Rhasspy](https://community.rhasspy.org/t/helper-library-to-develop-rhasspy-apps-in-python/969) we've been thinking about a framework for developing Rhasspy apps (or "skills", whatever one wants to call them). Rhasspy 2.5 will be using Hermes as its main protocol for communicating, both internally and externally (with skills/apps).

Since I like AppDaemon and use it for other purposes, I was thinking of implementing an [AppDaemon plugin](https://github.com/daniele-athome/hass-config/blob/master/custom_plugins/hermes/hermesplugin.py) (and [app base class](https://github.com/daniele-athome/hass-config/blob/master/custom_plugins/hermes/hermesapi.py)) for Hermes.

It's not mature yet, but would you be interested in integrating this into the project?
Also, would you accept PR for partial implementations? I mean the Hermes protocol has a lot of stuff, I might begin with something partial (but functioning, just not implementing the whole protocol) and then go on incrementally.

**Comment by Odianosen25** [comment]
@daniele-athome thanks for this. The Hermes protocol, isn’t it based on MQTT? So wouldn’t it be a function of using the MQTT plugin with some apps?

If it’s not possible, what will need fixing to enable the MQTT plugin to be used?

Regards

**Comment by daniele-athome** [comment]
It does use MQTT. As a matter of fact, my plugin [would be based on the MQTT plugin](https://github.com/daniele-athome/hass-config/blob/master/custom_plugins/hermes/hermesplugin.py#L14). It's just a way to specialize it so it can be simpler to use.

On top of that, I was thinking of implementing some annotations, similar to [what the guys at Rhasspy's are doing](https://github.com/rhasspy/rhasspy-hermes-app/blob/master/time_app.py).

**Comment by acockburn** [comment]
Hi Daniele - I think this is interesting!

If the MQTT plugin already does the basics, perhaps this would be a plugin that re-uses the MQTT plugin piece but substitutes a different API? That's not something we can to in AD yet but wouldn't be hard to fix up.

I don't mind if the implementation is initially incomplete as long as it is useful fro some purpose.

However we end up doing it - I see some value here and would be happy to add it to AD with the following caveats:

1. You maintain it
2. You document it

(This is simply because if I'm not using a thing I don't know enough to maintain it and don't have the bandwidth to take on any new stuff at the moment, although I might end up using it at some point!)

We would work with you to make sure you have the info you need, and any changes that might be needed to the AD core.

Will that work?

**Comment by daniele-athome** [comment]
Hey! Sorry for the delay, termination of the COVID lockdown had me busy on other fronts. I'm still using HASS/AppDaemon with this prototype so I'm still interested in seeing this part of AppDaemon.

> If the MQTT plugin already does the basics, perhaps this would be a plugin that re-uses the MQTT plugin piece but substitutes a different API? That's not something we can to in AD yet but wouldn't be hard to fix up.

I started writing just an API (i.e. the app base class) which used the MQTT plugin, but eventually it wasn't enough to provide an easy to use base (e.g. API classes can't publish events IIUC). So I also wrote the plugin by extending the MQTT plugin. It's a bit of hack in some places though (but not so much, feel free to review the current code if you'd like) because I had to tap into the MQTT plugin to do some stuff. But if I can do this in a more clean way I'm all ears.

> I don't mind if the implementation is initially incomplete as long as it is useful fro some purpose.

Good! I'll be focusing on basic functionality needed for creating Rhasspy "skills" (mainly responding to intents and voice interaction) and proceed from that gradually.

> However we end up doing it - I see some value here and would be happy to add it to AD with the following caveats:
> 1. You maintain it
> 2. You document it

I'm ok with that, since I'll be using it on a daily basis (my place is powered by this stuff, so I would be the first user of my own code). However I can't promise anything on the long term (as in more than a few years of maintenance, because you know life could happen :-), but as long as I'll be using it, I'll also maintain it - it would be in my own interest to do so.
